### PR TITLE
[IMP] *: check for validity of field parameters

### PR DESCRIPTION
**Pull request by xmo-odoo**
* Checks for the validity of parameter names to ward of typos (e.g. sting
  instead of string), long-outdated parameters (`type`) or parameters which don't
  make sense in context (e.g. `tracking` on non-mail models or `ondelete` on o2m)
* Fixes problematic cases related to the above
* Adds support for the `invisible` parameter which is in effective use

### CODE REVIEW

**Comment by xmo-odoo** [comment]
@rco-odoo hm it looks like ondelete does *something* for m2m records: the failure in products resolves if the ondelete=restrict is reintroduced (!!!)

**Comment by xmo-odoo** [comment]
Oh yeah the tests fail because ondelete *is* a valid attribute on m2m now.

File: addons/website/models/res_config_settings.py [hunk 1]
```diff
@@ -23,8 +23,7 @@ def _default_website(self):
     website_logo = fields.Binary(related='website_id.logo', readonly=False)
     language_ids = fields.Many2many(related='website_id.language_ids', relation='res.lang', readonly=False)
     website_language_count = fields.Integer(string='Number of languages', compute='_compute_website_language_count', readonly=True)
-    website_default_lang_id = fields.Many2one(string='Default language', related='website_id.default_lang_id',
-                                              readonly=False, relation='res.lang')
+    website_default_lang_id = fields.Many2one(string='Default language', related='website_id.default_lang_id', readonly=False)
```
Comment (rco-odoo) [inline addons/website/models/res_config_settings.py:26]: The alternative is to pass the comodel name as first positional argument
```suggestion
    website_default_lang_id = fields.Many2one('res.lang', string='Default language', related='website_id.default_lang_id', readonly=False)
```

File: addons/website/models/website.py [hunk 2]
```diff
@@ -105,7 +105,7 @@ def _default_logo(self):
     cdn_activated = fields.Boolean('Content Delivery Network (CDN)')
     cdn_url = fields.Char('CDN Base URL', default='')
     cdn_filters = fields.Text('CDN Filters', default=lambda s: '\n'.join(DEFAULT_CDN_FILTERS), help="URL matching those filters will be rewritten using the CDN Base URL")
-    partner_id = fields.Many2one(related='user_id.partner_id', relation='res.partner', string='Public Partner', readonly=False)
+    partner_id = fields.Many2one(related='user_id.partner_id', string='Public Partner', readonly=False)
```
Comment (rco-odoo) [inline addons/website/models/website.py:108]: Or as positional argument
```suggestion
    partner_id = fields.Many2one('res.partner', related='user_id.partner_id', string='Public Partner', readonly=False)
```

File: odoo/fields.py [hunk 3]
```diff
@@ -233,6 +232,7 @@ def _read_group_many2one_field(self, records, domain, order):
 
     string = None                       # field label
     help = None                         # field tooltip
+    invisible = False             # whether the field is invisible
```
Comment (rco-odoo) [inline odoo/fields.py:235]: Misalignment panic! :wink: 
```suggestion
    invisible = False                   # whether the field is invisible
```

File: odoo/fields.py [hunk 4]
```diff
@@ -345,19 +345,22 @@ def _get_attrs(self, model, name):
 
     def _setup_attrs(self, model, name):
         """ Initialize the field parameter attributes. """
+        # validate extra arguments
+        for key in self.args:
```
Comment (rco-odoo) [inline odoo/fields.py:349]: The field parameters are in `attrs` below, not `self.args`.

Actually, `self.args` will be empty most of the time here.  Consider a model definition like
```python
class Stuff(models.Model):
    _name = 'stuff'
    foo = fields.Char(stuff=42)    # Stuff.foo.args = {'stuff': 42}
```
The class put in the registry will actually inherit from `Stuff` and look like
```python
class ModelClass(Stuff, ...):
    _name = 'stuff'
    foo = fields.Char()    # ModelClass.foo.args = {}
```
The method `_setup_attrs` is actually invoked on `ModelClass.foo`, which introspects its class definition and determine `attrs` from `Stuff.foo.args` and `ModelClass.foo.args`.  The latter is non-empty only for automatic and custom fields.

Simply try the branch without the fixes, and you will see that it does not do what it is supposed to.

**Comment by rco-odoo** [comment]
I just checked performance during an install, the cost to validate is pretty small.

**Comment by rco-odoo** [comment]
Just fixed the error message.

**Comment by rco-odoo** [comment]
Voilà, just added a test.

**Comment by rco-odoo** [comment]
@robodoo rebase-ff r+

**Comment by MiquelRForgeFlow** [comment]
now maybe you could also check for company validity checks like done in https://github.com/odoo/odoo/pull/46721 🤓🙏

**Comment by rco-odoo** [comment]
@robodoo r+

**Comment by xmo-odoo** [comment]
@robodoo r+

**Comment by Feyensv** [comment]
@xmo-odoo Maybe the "m" commit https://github.com/odoo/odoo/pull/47917/commits/a42a97aaf54c352ba839d8d6223692ba0585377b shouldn't be merged like this ;) (the merging failed anyway)

File: addons/account/models/account_move.py [hunk 5]
```diff
@@ -3578,6 +3578,10 @@ def field_will_change(line, field_name):
 
         return result
 
+    def _valid_field_parameter(self, field, name):
+        # I can't even
```
Comment (Feyensv) [inline addons/account/models/account_move.py:3582]: ?

Comment (xmo-odoo) [reply addons/account/models/account_move.py:3582]: https://knowyourmeme.com/memes/i-can-t-even

**Comment by xmo-odoo** [comment]
@robodoo r+

**Comment by william-andre** [comment]
This is great, but it doesn't seem to work anymore.
I could make a declaration 
```python
    amount_residual = fields.Monetary(string='Amount Due', store=True, something=True,
        compute='_compute_amount')
```
without any warnings.
I also tried with cases that seem to have been detected by this PR

It seems that the parameters are not forwarded until the point of the test for invalid keys. (from a quick search, they seem to be removed at `_setup_related_full`)

Maybe adding a test would be a good idea?

**Comment by xmo-odoo** [comment]
> It seems that the parameters are not forwarded until the point of the test for invalid keys. (from a quick search, they seem to be removed at `_setup_related_full`)

As the name indicates, `_setup_related_full` is only called to setup related fields. 

The issue seems to be that the fields [are copied without keeping any of the attributes](https://github.com/odoo/odoo/blob/3e91fa1dbf9056f1557ff7a747ba3407e851b18e/odoo/models.py#L2731), and since that code didn't change recently, we've recently changed the codepath we take or -- much more likely -- calls to `_setup_base` were optimised away to only be performed on the final / leaf classes. @rco-odoo might be able to tell.

**Comment by xmo-odoo** [comment]
Issue might be due to f694caac416cd66c3878cf6352855cb9765c3b45 as it seems to move around calls to `_add_field` (which calls `setup_base`).

**Pull request closed without merging**